Enable deposit recovery from any state

ffakenz opened the pull request:
<!-- Describe your change here -->

🔑 **Motivation** 
> Closes #1812

Pending deposits are currently only tracked in the `Open` state.  

This prevents clients from recovering funds through their running node once the head transitions into later states such as `Closed` or after `Fanout`.

As a result, users are forced to build and submit recovery transactions manually, which is error-prone and inconvenient.

To improve usability, deposit recovery should be possible through the API regardless of the current state of the Hydra head.

🔄 **Breaking Changes**

* `HeadState` no longer tracks pending deposits or the current chain slot.  
  These are now part of the newly introduced node-level `NodeState`.  
  This refactor also opens the door to moving `ChainState` out of `HeadState`, as noted in [this TODO](https://github.com/cardano-scaling/hydra/blob/3dffe727ea31d908e4516f37447385fe50051578/hydra-node/src/Hydra/HeadLogic/State.hs#L39).

* The `Checkpoint` event, and consequently the `EventLogRotated` server output, now carry the full `NodeState` instead of just the `HeadState`.  
  This enables clients (e.g. the TUI) to fully recover after event-log rotation.

📝 **Notes**

* Deposits are still only **recorded, activated, and expired** while the head is `Open`, so recovery is limited to deposits observed in previously opened heads.

* `HeadLogic` now aggregates events in two phases:
  1. Head-level  
  2. Node-level

  > _This introduces a dependency between the two update steps: the correctness of the node-level state update relies on the head-level state update. Any bug or inconsistency in the head-level update will cascade into the node-level view._


---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

ffakenz committed:
draft base e2e spec

ffakenz committed:
rename NodeState type into NodeStateHandler

ffakenz committed:
draft data type changes of moving things to node-level

ffakenz committed:
move current slot to node-level

> this allows handling client recover in all head states

ffakenz committed:
update CHANGELOG

ffakenz committed:
fix tui

ffakenz committed:
handle recover tx at node-level

ffakenz committed:
remove spec ignoring recoverTx of another head

> as we moved this observation to node-level

noonio commented on hydra-node/src/Hydra/HeadLogic/Outcome.hs:
```diff
@@ -145,13 +145,13 @@ data StateChanged tx
   | TxInvalid {headId :: HeadId, utxo :: UTxOType tx, transaction :: tx, validationError :: ValidationError}
   | LocalStateCleared {headId :: HeadId, snapshotNumber :: SnapshotNumber}
-  | Checkpoint {state :: HeadState tx}
+  | Checkpoint {state :: NodeState tx}
```
I'm not sure how I feel about this.

Why did we make the Checkpoint over the node-state instead of the head state? Checkpoints seem more related to a specific head?

ffakenz replied:
Because the TUI keeps track of pendingDeposits, and during restart it relies on them to recover its state, especially when the node has rotated.

noonio replied:
Okay; I can live with this I think.

Maybe it's a naming thing; but it feels unfortunate to have "node" things in the file related to HeadLogic Outcomes; but I can see your point.

noonio commented on hydra-node/src/Hydra/HeadLogic/State.hs:
```diff
@@ -25,6 +25,27 @@ import Hydra.Tx.Snapshot (
+  , pendingDeposits :: PendingDeposits tx
+  -- ^ Pending deposits as observed on chain.
+  -- TODO: could even move the chain state here (also see todo below)
+  -- , chainState :: ChaiStateType tx
```
```suggestion
  -- , chainState :: ChainStateType tx
```

noonio commented on hydra-node/src/Hydra/HeadLogic/State.hs:
```diff
@@ -25,6 +25,27 @@ import Hydra.Tx.Snapshot (
+  { headState :: HeadState tx
+  , pendingDeposits :: PendingDeposits tx
+  -- ^ Pending deposits as observed on chain.
+  -- TODO: could even move the chain state here (also see todo below)
```
I don't think there's a `TODO` that this is referencing?

ffakenz replied:
yes [this TODO](https://github.com/cardano-scaling/hydra/blob/3dffe727ea31d908e4516f37447385fe50051578/hydra-node/src/Hydra/HeadLogic/State.hs#L39).

noonio commented on hydra-node/test/Hydra/Events/RotationSpec.hs:
```diff
@@ -44,7 +44,7 @@ spec = parallel $ do
           -- this is hardcoded to ensure we get a checkpoint + a single event at the end
           let rotationConfig = RotateAfter (Positive 3)
-          let s0 = Idle IdleState{chainState = SimpleChainState{slot = ChainSlot 0}}
+          let s0 = NodeState{headState = Idle IdleState{chainState = SimpleChainState{slot = ChainSlot 0}}, pendingDeposits = mempty, currentSlot = ChainSlot 0}
```
I see lots of these new initialisations; maybe could have `emptyNodeState { ... }` to override only the fields we care about.

noonio commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -1335,7 +1335,7 @@ createHydraNode tracer ledger chainState signingKey otherParties outputs message
         , env
         , ledger
-        , nodeState
+        , nodeState = nodeStateHandler
```
I feel like to avoid confusion this field should be called `nodeStateHandler`.

noonio commented on CHANGELOG.md:
```diff
@@ -41,6 +41,8 @@ changes.
 
 - Label threads, queues and vars.
 
+- **BREAKING** Enable handling client recover in all head states.
```
Please can you link the issue in the changelog entry?

noonio commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1432,6 +1434,56 @@ update env ledger st ev = case (st, ev) of
+      ns@NodeState{headState = st, pendingDeposits} = nodeState{headState = headState'}
+   in case sc of
+        HeadOpened{chainState} ->
+          ns{pendingDeposits = mempty, currentSlot = chainStateSlot chainState}
```
I don't think this is right; is it? When it observes a head open it clears the pending deposits?

ffakenz replied:
good catch!

noonio replied:
Let me note explicitly that it'd be good to get this covered by a test :)

noonio commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1432,6 +1434,56 @@ update env ledger st ev = case (st, ev) of
+aggregateNodeState :: IsChainState tx => NodeState tx -> StateChanged tx -> NodeState tx
+aggregateNodeState nodeState@NodeState{headState} sc =
+  let headState' = aggregate headState sc
+      ns@NodeState{headState = st, pendingDeposits} = nodeState{headState = headState'}
```
What is happending here? We already have `nodeState@NodeState{headState}` two lines above? I think these lines can be combined a bit to be less confusing.

ffakenz replied:
Agree, destructuring twice is confusing. I’ll look into simplifying this so it’s clearer.

noonio commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1432,6 +1434,56 @@ update env ledger st ev = case (st, ev) of
+                    newUTxO = maybe mempty (\Deposit{deposited} -> deposited) deposit
+                 in ns
+                      { headState =
+                          Open
```
If this is headState changes why isn't it deferred to the aggregateHeadState function?

ffakenz replied:
Because in this case the HeadState changes actually depend on pendingDeposits:
localUTxO appends new UTxO that comes from a deposit we are deleting from pending.

The only parts that are truly head-specific and independent from deposits are version and currentDepositTxId.

So in summary, I decided to handle this event here since it feels more related to deposits than to the head itself. Does that make sense?

Would you prefer splitting the handling of version and currentDepositTxId into the head-state aggregation instead?

> Note: one alternative could be to pass pendingDeposits into the head-state aggregation, but that would flip the aggregation order and introduce a tighter coupling between the two aggregation steps.

Looking forward to your thoughts

noonio commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1622,84 +1672,12 @@ aggregate st = \case
-       where
-        CoordinatedHeadState{pendingDeposits} = coordinatedHeadState
-    _otherState -> st
-  CommitFinalized{chainState, newVersion, depositTxId} ->
```
I don't see why this one was deleted?

I guess because it deletes it from pendingDeposits; but maybe the headState change should be left here and the pendingDeposits change brought into the other one?

ffakenz replied:
Related to the above https://github.com/cardano-scaling/hydra/pull/2217/files#r2322216524

noonio reviewed (changes requested):
Looking good @ffakenz thanks!

I left a few questions and comments that I'd like your opinion on.

ffakenz committed:
link issue and pr number to changelog entry

ffakenz committed:
add helper to init node state

> also rename DraftHydraNode and HydraNode nodeState field no nodeStateHandler to avoid confusion

ffakenz committed:
simplify node state aggregation by avoiding double destructuring

ffakenz committed:
reproduce issue where we can't recover deposit from previous open head

ffakenz committed:
fix: can recover deposit from previous open head

ffakenz committed:
refactor e2e scenario

noonio reviewed (approved):
Nice

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -976,6 +979,8 @@ onOpenChainTick env st chainTime =
   plusTime = flip addUTCTime
 
+  -- REVIEW! check what if there are more than 1 new active deposit
+  -- What is the sorting criteria to pick next?
```
Could you answer this question?

ffakenz replied:
Currently it picks the first active deposit based on the `Ord` instance of `TxIdType`.

In this [PR](https://github.com/cardano-scaling/hydra/pull/2263) I propose changing this to FIFO, using the `created` field, so deposits are processed in arrival order.

wdyt?

ch1bo commented on CHANGELOG.md:
```diff
@@ -39,6 +39,9 @@ changes.
 
 - Label threads, queues and vars.
 
+- **BREAKING** Enable handling client recover in all head states.
```
How exactly is this a breaking change? I suspect it's because of the `Checkpoint` event format change, but we should try to be very clear what this means to a users. The changelog is a communication channel between developers and users (people running `hydra-node`).

ch1bo commented on hydra-node/test/Hydra/API/HTTPServerSpec.hs:
```diff
@@ -69,6 +70,7 @@ spec = do
     roundtripAndGoldenSpecs (Proxy @(ReasonablySized TransactionSubmitted))
     roundtripAndGoldenSpecs (Proxy @(ReasonablySized (SideLoadSnapshotRequest Tx)))
     roundtripAndGoldenSpecs (Proxy @(ReasonablySized (HeadState Tx)))
+    roundtripAndGoldenSpecs (Proxy @(ReasonablySized (NodeState Tx)))
```
Why do you roundtrip test this here? It is not exposed through the HTTP API (right now) so we do not need to detect changes to it (here)

ffakenz replied:
Because It is actually exposed through the WS API  as part of the EventLogRotated.

ch1bo commented on hydra-node/src/Hydra/Node.hs:
```diff
@@ -404,27 +404,29 @@ processEffects node tracer inputId effects = do
 
 -- ** Manage state
 
+-- TODO! pendingDeposits :: Map (TxIdType tx) (Deposit tx)
```
I don't see what is todo about this? remove this comment

ch1bo commented on hydra-node/src/Hydra/HeadLogic/State.hs:
```diff
@@ -25,6 +25,35 @@ import Hydra.Tx.Snapshot (
 
+type PendingDeposits tx = Map (TxIdType tx) (Deposit tx)
+
+-- FIXME: move to a dedicated module (maybe with deposits too?)
```
@ffakenz you forgot to address this FIXME or it is rather a TODO (which you might have wanted to do while working on this?)

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1370,12 +1374,10 @@ update env ledger st ev = case (st, ev) of
+    onOpenNetworkReqDec env ledger ttl currentSlot openState transaction
   (Open OpenState{headId = ourHeadId}, ChainInput Observation{observedTx = OnDepositTx{headId, depositTxId, deposited, created, deadline}, newChainState})
     | ourHeadId == headId ->
         newState DepositRecorded{chainState = newChainState, headId, depositTxId, deposited, created, deadline}
```
Should: handle Deposit events independently of `HeadState`

While the scope of this PR was to allow recovers in any state (originally the issue was only about closed state), I wondered why we did not generalize here too. It should be quite straight-forward to do this too?

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
Do the changes make this named `NodeLogic` or just `Logic`? I saw the composition on the `aggregate` function where the node-level calls the head-level, could/should we do the same for the `update` side of things?

Just some thoughts in which direction this could be refactored to become more clean for the next developer touching this.

ffakenz replied:
From my POV, `NodeLogic` feels like the place where the `stepHydraNode` function should live, which is currently the `Hydra.Node` module.

`Hydra.HeadLogic` handles the off-chain protocol, and even though deposits are now resolved at the node level, they’re still part of that protocol layer.

That said, I think it’s fine as is and wouldn’t change it.  
However, your question is totally valid and worth raising with the team.

noonio replied:
I think @ch1bo is raising a similar feeling i had; that it's a bit unfortunate that "node"-level logic is sneaking into this file.

I think part of the trouble is that the deposit's cut across both the node and the Head; so they kind of need to be considered at the same time in that sense.

I'm not entirely sure what the answer is.

I think, as usual, a little sketch of the relevant logic in Miro, with clear boundaries drawn, might be illuminating :)

@ffakenz would you be willing to do that drawing and we can discuss from there?

ffakenz replied:
<img width="905" height="562" alt="image" src="https://github.com/user-attachments/assets/f2f1fe1d-b334-4eb6-b5ee-b9e4e725482a" />

ch1bo reviewed (commented):
As we talked about the `NodeState` a couple weeks ago and drafted the types together I thought I would have a look at how you ended up doing things.

Good job overall, but I feel a few things had been left behind:

ffakenz commented:
> As we talked about the `NodeState` a couple weeks ago and drafted the types together I thought I would have a look at how you ended up doing things.
> 
> Good job overall, but I feel a few things had been left behind:

@ch1bo thank you very much for your review and feedback.

I've created a follow-up [PR](https://github.com/cardano-scaling/hydra/pull/2256) to address them.
